parser-util/main.cc

```diff
@@ -0,0 +1,346 @@
+/* ========================================================================== *
+ *
+ * Parse various URIs and junk using `nix' libraries and expose them in a
+ * trivially simple way so that they can be used consumed by other software.
+ *
+ * -------------------------------------------------------------------------- */
+
+#include <stddef.h>
+#include <iostream>
+#include <string>
+#include <cstring>
+#include <nlohmann/json.hpp>
+#include <nix/shared.hh>
+#include <nix/eval.hh>
+#include <nix/eval-inline.hh>
+#include <nix/flake/flake.hh>
+#include <nix/store-api.hh>
+#include <cassert>
+#include <ranges>
+
+
+/* -------------------------------------------------------------------------- */
+
+static nix::flake::LockFlags floxFlakeLockFlags = {
+  .updateLockFile = false
+, .writeLockFile  = false
+, .applyNixConfig = false
+};
+
+
+/* -------------------------------------------------------------------------- */
+
+namespace nix {
+    void
+  to_json( nlohmann::json & j, const FlakeRef & ref )
+  {
+    j = { { "string", ref.to_string() }
+        , { "attrs",  fetchers::attrsToJSON( ref.toAttrs() ) }
+        };
+  }
+}  /* End namespace `nix' */
+
+
+/* -------------------------------------------------------------------------- */
+
+  static nlohmann::json
+parseURI( const char * arg )
+{
+  try
+    {
+      nix::ParsedURL       url    = nix::parseURL( arg );
+      nix::ParsedUrlScheme scheme = nix::parseUrlScheme( url.scheme );
+
+      nlohmann::json s = {
```

Comment:
Let's give this a more descriptive name than `s`. I'm not sure what `s` is supposed to stand for here.